File: src/services/ssh.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, UnlistenFn } from "@tauri-apps/api/event";

export interface JumpHostConnect {
  host: string;
  port: number;
  username: string;
  password?: string;
  privateKey?: string;
  passphrase?: string;
}

export async function sshConnect(params: {
  sessionId: string;
  host: string;
  port: number;
  username: string;
  password?: string;
  privateKey?: string;
  passphrase?: string;
  connectionId?: string;
  jumpHosts?: JumpHostConnect[];
  envVars?: [string, string][];
  agentForwarding?: boolean;
  legacyAlgorithms?: boolean;
  preCommand?: string;
  autoForward?: boolean;
  shellIntegration?: boolean;
  keepaliveIntervalSecs: number;
  keepaliveMax: number;
  persist?: boolean;
  restore?: boolean;
  /** Re-attach/join an existing multiplexer session; never creates one.
   * Rejects with SESSION_ENDED when the session is gone on the host. */
  attachOnly?: boolean;
  cols?: number;
  rows?: number;
  /** Directory the shell starts in. First connect only — a reconnect must not
   * move a session the user has since navigated elsewhere. */
  initialCwd?: string;
}): Promise<void> {
  return invoke("ssh_connect", {
    sessionId: params.sessionId,
    host: params.host,
    port: params.port,
    username: params.username,
    password: params.password ?? null,
    privateKey: params.privateKey ?? null,
    passphrase: params.passphrase ?? null,
    connectionId: params.connectionId ?? null,
    jumpHosts: params.jumpHosts && params.jumpHosts.length > 0 ? params.jumpHosts : null,
    envVars: params.envVars && params.envVars.length > 0 ? params.envVars : null,
    agentForwarding: params.agentForwarding ?? false,
    legacyAlgorithms: params.legacyAlgorithms ?? false,
    preCommand: params.preCommand ?? null,
    autoForward: params.autoForward ?? true,
    shellIntegration: params.shellIntegration ?? null,
    keepaliveIntervalSecs: params.keepaliveIntervalSecs,
    keepaliveMax: params.keepaliveMax,
    persist: params.persist ?? null,
    restore: params.restore ?? null,
    attachOnly: params.attachOnly ?? null,
    cols: params.cols ?? null,
    rows: params.rows ?? null,
    initialCwd: params.initialCwd ?? null,
  });
}

/** Resolves to whether the persistent multiplexer session was confirmed
 * killed on the host (shared sessions: kill happens only when no other
 * client/device uses it; `attached` = our own channel is still attached). */
export async function sshDisconnect(
  sessionId: string,
  postCommand?: string,
  killPersistent?: boolean,
  attached?: boolean,
  reconnecting?: boolean,
): Promise<boolean> {
  return invoke("ssh_disconnect", {
    sessionId,
    postCommand: postCommand ?? null,
    killPersistent: killPersistent ?? null,
    attached: attached ?? null,
    reconnecting: reconnecting ?? null,
  });
}

/** Drop the SSH connection as the first half of a reconnect under the same
 * session id. Keeps the session's port forwards up: they are re-armed by the
 * matching `sshConnect`, so ad-hoc tunnels survive the round trip. */
export async function sshDisconnectForReconnect(sessionId: string): Promise<void> {
  await sshDisconnect(sessionId, undefined, undefined, undefined, true).catch(() => {});
}

export async function sshSendInput(sessionId: string, data: Uint8Array): Promise<void> {
  return invoke("ssh_send_input", { sessionId, data: Array.from(data) });
}

export async function sshResize(sessionId: string, cols: number, rows: number): Promise<void> {
  return invoke("ssh_resize", { sessionId, cols, rows });
}

export async function sshDetectDistro(sessionId: string): Promise<string> {
  return invoke("ssh_detect_distro", { sessionId });
}

export interface SystemInfo {
  pretty_name: string;
  version_id: string;
  kernel: string;
  arch: string;
}

export async function sshGetSystemInfo(sessionId: string): Promise<SystemInfo> {
  return invoke("ssh_get_system_info", { sessionId });
}

export interface SshExecResult {
  stdout: string;
  stderr: string;
  /** null when the channel closed without ever reporting an exit status — treat as a failure, not as 0. */
  exit_code: number | null;
}

export async function sshExecCommand(params: {
  host: string;
  port: number;
  username: string;
  password?: string;
  privateKey?: string;
  passphrase?: string;
  command: string;
  legacyAlgorithms?: boolean;
}): Promise<SshExecResult> {
  return invoke("ssh_exec_command", {
    host: params.host,
    port: params.port,
    username: params.username,
    password: params.password ?? null,
    privateKey: params.privateKey ?? null,
    passphrase: params.passphrase ?? null,
    command: params.command,
    legacyAlgorithms: params.legacyAlgorithms ?? false,
  });
}

export async function sshKillPersistent(params: {
  host: string;
  port: number;
  username: string;
  password?: string;
  privateKey?: string;
  passphrase?: string;
  sessionId: string;
  legacyAlgorithms?: boolean;
}): Promise<boolean> {
  return invoke("ssh_kill_persistent", {
    host: params.host,
    port: params.port,
    username: params.username,
    password: params.password ?? null,
    privateKey: params.privateKey ?? null,
    passphrase: params.passphrase ?? null,
    sessionId: params.sessionId,
    legacyAlgorithms: params.legacyAlgorithms ?? false,
  });
}

export async function onSshOutput(
  sessionId: string,
  callback: (data: Uint8Array) => void,
): Promise<UnlistenFn> {
  return listen<number[]>(`ssh-output-${sessionId}`, (event) => {
    callback(new Uint8Array(event.payload));
  });
}

/** `remoteExit` is true when the far side sent an exit-status/exit-signal
 * before closing — the remote command ended on its own, it was not a drop. */
export async function onSshClosed(
  sessionId: string,
  callback: (remoteExit: boolean) => void,
): Promise<UnlistenFn> {
  return listen<boolean>(`ssh-closed-${sessionId}`, (event) => {
    callback(event.payload === true);
  });
}

// Backend-polled cwd for persistent sessions, where the tmux/screen multiplexer
// swallows the shell's OSC 7 so the terminal's own OSC 7 handler never sees it.
export async function onSshCwd(
  sessionId: string,
  callback: (cwd: string) => void,
): Promise<UnlistenFn> {
  return listen<string>(`ssh-cwd-${sessionId}`, (event) => {
    callback(event.payload);
  });
}

File: src-tauri/src/commands/ssh.rs
use crate::known_hosts::{KnownHostsStore, PendingConflicts};
use crate::port_forward::PortForwardManager;
use crate::ssh::{
    client::{self, JumpHostConnect},
    session::SessionManager,
};
use std::sync::Arc;
use tauri::AppHandle;

#[tauri::command]
pub async fn ssh_connect(
    app: AppHandle,
    state: tauri::State<'_, SessionManager>,
    pf: tauri::State<'_, PortForwardManager>,
    known_hosts: tauri::State<'_, Arc<KnownHostsStore>>,
    pending_conflicts: tauri::State<'_, Arc<PendingConflicts>>,
    session_id: String,
    host: String,
    port: u16,
    username: String,
    password: Option<String>,
    private_key: Option<String>,
    passphrase: Option<String>,
    connection_id: Option<String>,
    jump_hosts: Option<Vec<JumpHostConnect>>,
    env_vars: Option<Vec<(String, String)>>,
    agent_forwarding: bool,
    pre_command: Option<String>,
    auto_forward: bool,
    shell_integration: Option<bool>,
    keepalive_interval_secs: Option<u64>,
    keepalive_max: Option<usize>,
    persist: Option<bool>,
    restore: Option<bool>,
    attach_only: Option<bool>,
    cols: Option<u32>,
    rows: Option<u32>,
    legacy_algorithms: Option<bool>,
    initial_cwd: Option<String>,
) -> Result<(), String> {
    let connected = client::connect(
        app,
        session_id.clone(),
        &host,
        port,
        &username,
        password.as_deref(),
        private_key.as_deref(),
        passphrase.as_deref(),
        jump_hosts.unwrap_or_default(),
        env_vars.unwrap_or_default(),
        agent_forwarding,
        pre_command,
        shell_integration.unwrap_or(true),
        Arc::clone(&*known_hosts),
        Arc::clone(&*pending_conflicts),
        keepalive_interval_secs.unwrap_or(3),
        keepalive_max.unwrap_or(3),
        persist.unwrap_or(true),
        restore.unwrap_or(false),
        attach_only.unwrap_or(false),
        cols.filter(|c| *c > 0).unwrap_or(80),
        rows.filter(|r| *r > 0).unwrap_or(24),
        legacy_algorithms.unwrap_or(false),
        initial_cwd,
    )
    .await?;

    state.add(session_id.clone(), connected).await;

    // Register this terminal under its host key so port forwarding is shared
    // across all terminals of the same connection (not duplicated per terminal).
    let cid = connection_id.as_deref().unwrap_or("");
    pf.register_session(&session_id, cid).await;

    if let Ok(handle) = state.get_session_handle(&session_id).await {
        let routes = state
            .get_remote_routes(&session_id)
            .await
            .unwrap_or_else(|_| {
                Arc::new(tokio::sync::Mutex::new(std::collections::HashMap::new()))
            });
        // No-op for a first connect (nothing to re-arm) and for a reconnect with
        // only local/dynamic tunnels, which follow the new handle by themselves.
        pf.rearm_after_reconnect(&session_id, Arc::clone(&handle), Arc::clone(&routes))
            .await;
        pf.auto_activate_rules(&session_id, cid, Arc::clone(&handle), routes)
            .await;
        let _ = pf.set_auto_detect(&session_id, auto_forward, handle).await;
    }

    Ok(())
}

#[tauri::command]
pub async fn ssh_disconnect(
    state: tauri::State<'_, SessionManager>,
    pf: tauri::State<'_, PortForwardManager>,
    session_id: String,
    post_command: Option<String>,
    kill_persistent: Option<bool>,
    attached: Option<bool>,
    reconnecting: Option<bool>,
) -> Result<bool, String> {
    // A reconnect drops the SSH connection and immediately dials again under the
    // same session id, so the terminal is not leaving its host — the forwards
    // stay up and `ssh_connect` re-arms them. Tearing them down here is what used
    // to lose every ad-hoc tunnel across a reconnect, since only rule-backed ones
    // were rebuilt afterwards.
    if !reconnecting.unwrap_or(false) {
        // Host-scoped port forwarding: only tear down when the last terminal of the
        // host closes. If the terminal that owned the forwards leaves while siblings
        // remain, hand the forwards off to a surviving terminal's SSH handle so they
        // keep working.
        let (key, remaining, was_owner) = pf.detach_session(&session_id).await;
        if remaining.is_empty() {
            pf.teardown_key(&key).await;
        } else {
            if was_owner {
                if let Some(survivor) = remaining.first() {
                    if let Ok(handle) = state.get_session_handle(survivor).await {
                        let routes = state.get_remote_routes(survivor).await.unwrap_or_else(|_| {
                            Arc::new(tokio::sync::Mutex::new(std::collections::HashMap::new()))
                        });
                        pf.rebind_to_handle(&key, survivor, handle, routes).await;
                    }
                }
            }
            pf.clear_session_panel(&session_id).await;
        }
    }

    state
        .disconnect(
            &session_id,
            post_command,
            kill_persistent.unwrap_or(false),
            attached.unwrap_or(false),
        )
        .await
}

#[tauri::command]
pub async fn ssh_send_input(
    state: tauri::State<'_, SessionManager>,
    session_id: String,
    data: Vec<u8>,
) -> Result<(), String> {
    state.send_data(&session_id, &data).await
}

#[tauri::command]
pub async fn ssh_detect_distro(
    state: tauri::State<'_, SessionManager>,
    session_id: String,
) -> Result<String, String> {
    state.detect_distro(&session_id).await
}

#[tauri::command]
pub async fn ssh_get_system_info(
    state: tauri::State<'_, SessionManager>,
    session_id: String,
) -> Result<crate::ssh::session::SystemInfo, String> {
    state.get_system_info(&session_id).await
}

#[tauri::command]
pub async fn ssh_resize(
    state: tauri::State<'_, SessionManager>,
    session_id: String,
    cols: u32,
    rows: u32,
) -> Result<(), String> {
    state.resize(&session_id, cols, rows).await
}

#[derive(serde::Serialize)]
pub struct SshExecResult {
    pub stdout: String,
    pub stderr: String,
    /// None when the channel closed without ever sending an exit status —
    /// distinct from `Some(0)`, so a caller can't mistake "unknown" for "ok".
    pub exit_code: Option<i32>,
}

#[tauri::command]
pub async fn ssh_exec_command(
    known_hosts: tauri::State<'_, Arc<KnownHostsStore>>,
    host: String,
    port: u16,
    username: String,
    password: Option<String>,
    private_key: Option<String>,
    passphrase: Option<String>,
    command: String,
    legacy_algorithms: Option<bool>,
) -> Result<SshExecResult, String> {
    use tokio::time::{timeout, Duration};

    let handle = client::connect_authenticated(
        Arc::clone(&*known_hosts),
        &host,
        port,
        &username,
        password.as_deref(),
        private_key.as_deref(),
        passphrase.as_deref(),
        legacy_algorithms.unwrap_or(false),
    )
    .await?;

    let mut channel = handle
        .channel_open_session()
        .await
        .map_err(|e| format!("Channel error: {}", e))?;

    channel
        .exec(true, command.as_str())
        .await
        .map_err(|e| format!("Exec error: {}", e))?;

    let mut stdout = Vec::new();
    let mut stderr = Vec::new();
    let mut exit_code: Option<i32> = None;

    // Read to Close rather than Eof: the exit status usually follows Eof, and
    // breaking there is what loses it.
    let timed_out = timeout(Duration::from_secs(30), async {
        while let Some(msg) = channel.wait().await {
            match msg {
                russh::ChannelMsg::Data { data } => stdout.extend_from_slice(&data),
                russh::ChannelMsg::ExtendedData { data, .. } => stderr.extend_from_slice(&data),
                russh::ChannelMsg::ExitStatus { exit_status } => {
                    exit_code = Some(exit_status as i32)
                }
                russh::ChannelMsg::Close => break,
                _ => {}
            }
        }
    })
    .await
    .is_err();

    let _ = handle
        .disconnect(russh::Disconnect::ByApplication, "Done", "en")
        .await;

    // A hard 30s timeout means the remote command never finished: "exit 0,
    // partial stdout" would misreport that as a success.
    if timed_out {
        return Err("Remote command timed out after 30s".to_string());
    }

    Ok(SshExecResult {
        stdout: String::from_utf8_lossy(&stdout).to_string(),
        stderr: String::from_utf8_lossy(&stderr).to_string(),
        exit_code,
    })
}

#[tauri::command]
pub async fn ssh_kill_persistent(
    known_hosts: tauri::State<'_, Arc<KnownHostsStore>>,
    host: String,
    port: u16,
    username: String,
    password: Option<String>,
    private_key: Option<String>,
    passphrase: Option<String>,
    session_id: String,
    legacy_algorithms: Option<bool>,
) -> Result<bool, String> {
    use tokio::io::AsyncReadExt;
    use tokio::time::{timeout, Duration};

    let handle = client::connect_authenticated(
        Arc::clone(&*known_hosts),
        &host,
        port,
        &username,
        password.as_deref(),
        private_key.as_deref(),
        passphrase.as_deref(),
        legacy_algorithms.unwrap_or(false),
    )
    .await?;

    let command = crate::shell_integration::force_kill_command(&session_id);

    let channel = handle
        .channel_open_session()
        .await
        .map_err(|e| format!("Channel error: {}", e))?;
    channel
        .exec(true, command.as_str())
        .await
        .map_err(|e| format!("Exec error: {}", e))?;

    let mut stream = channel.into_stream();
    let mut output = Vec::new();
    let _ = timeout(Duration::from_secs(15), async {
        let mut buf = [0u8; 256];
        loop {
            match stream.read(&mut buf).await {
                Ok(0) | Err(_) => break,
                Ok(n) => output.extend_from_slice(&buf[..n]),
            }
        }
    })
    .await;

    let _ = handle
        .disconnect(russh::Disconnect::ByApplication, "Done", "en")
        .await;

    Ok(String::from_utf8_lossy(&output).contains("VOLTIUS_KILLED"))
}
